src/Type/Php/ArrayMergeFunctionDynamicReturnTypeExtension.php

```diff
@@ -138,4 +135,29 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 		return $arrayType;
 	}
 
+	/**
+	 * array{0: 17, foo: 'bar'}|array{0: 19, foo: 'bar', fofoo: 'barbar'}
+	 * ->
+	 * array{0: 17|19, foo: 'bar', foofo?: 'barbar'}
+	 */
+	private function untagConstantArrayUnion(Type $constantArrayType): Type
+	{
+		$constantArrayTypes = $constantArrayType->getConstantArrays();
+		if (count($constantArrayTypes) === 1) {
+			return $constantArrayTypes[0];
+		}
+
+		$builder = ConstantArrayTypeBuilder::createEmpty();
+		$keyTypes = $constantArrayType->getKeysArray()->getIterableValueType()->getFiniteTypes();
+		foreach ($keyTypes as $keyType) {
+			$builder->setOffsetValueType(
+				$keyType,
+				$constantArrayType->getOffsetValueType($keyType),
+				!$constantArrayType->hasOffsetValueType($keyType)->yes(),
+			);
+		}
+
+		return $builder->getArray();
+	}
```

Comment:
took me a bit to come up with this and I think it's not bad, but it still feels slightly strange to do that. is there maybe a better or more generic way? :D